src/lib/replication/repl_dev/raft_state_machine.cpp

```diff
@@ -0,0 +1,510 @@
+#include <iomgr/iomgr_timer.hpp>
+#include <sisl/logging/logging.h>
+#include <sisl/fds/utils.hpp>
+#include <sisl/fds/vector_pool.hpp>
+
+#include "repl_dev/raft_state_machine.h"
+#include "repl_dev/raft_repl_dev.h"
+
+SISL_LOGGING_DECL(replication)
+
+namespace homestore {
+
+RaftStateMachine::RaftStateMachine(RaftReplDev& rd) : m_rd{rd}, m_group_id{m_rd.group_id()} {
+    m_success_ptr = nuraft::buffer::alloc(sizeof(int));
+    m_success_ptr->put(0);
+}
+
+raft_buf_ptr_t RaftStateMachine::pre_commit_ext(nuraft::state_machine::ext_op_params const& params) {
+    // Leader precommit is processed in next callback, since lsn would not have been known to repl layer till we get
+    // the next callback.
+    if (!m_rd.is_leader()) {
+        int64_t lsn = s_cast< int64_t >(params.log_idx);
+        raft_buf_ptr_t data = params.data;
+
+        RD_LOG(DEBUG, "pre_commit: {}, size: {}", lsn, data->size());
+        repl_req_ptr_t rreq = lsn_to_req(lsn);
+
+        m_rd.m_listener->on_pre_commit(rreq->lsn, rreq->header, rreq->key, rreq);
+    }
+    return m_success_ptr;
+}
+
+void RaftStateMachine::after_precommit_in_leader(nuraft::raft_server::req_ext_cb_params const& params) {
+    repl_req_ptr_t rreq = repl_req_ptr_t(r_cast< repl_req_ctx* >(params.context));
+    link_lsn_to_req(rreq, int64_cast(params.log_idx));
+    m_rd.m_listener->on_pre_commit(rreq->lsn, rreq->header, rreq->key, rreq);
+}
+
+raft_buf_ptr_t RaftStateMachine::commit_ext(nuraft::state_machine::ext_op_params const& params) {
+    int64_t lsn = s_cast< int64_t >(params.log_idx);
+    raft_buf_ptr_t data = params.data;
+
+    RD_LOG(DEBUG, "apply_commit: {}, size: {}", lsn, data->size());
+
+    repl_req_ptr_t rreq = lsn_to_req(lsn);
+    if (m_rd.is_leader()) {
+        // This is the time to ensure flushing of journal happens in leader
+        if (m_rd.m_data_journal->last_durable_index() < uint64_cast(lsn)) { m_rd.m_data_journal->flush(); }
```

Comment:
I am little confused about this behaviors:

1. will journal log flush only happens when it can be commited? 
2. why only happens in leader side? follower need flush the journal log too, right?

Reply:
Few points to note:
The data_journal->flush() call here ensures that logs up to now has been flushed. But it might have been already flushed because of timer or other group commit, in that case will not do anything. In general you can call this method as late as possible.

To your question:
1. Leader will need to ensure flushing only after the Raft confirms it is committed. Until that time, it is not required to be flushed, because after we commit, we need to record the commit lsn in the superblk (in-memory initially), before we respond to the user. So it is right to flush here.

2. Follower will flush on_pre_commit itself. Without flushing if it reports to the leader that it is precommitted and then crashed (and it happened on say all replicas), leader will go ahead and commit and respond to user, only later that data is lost. So we must flush during on_pre_commit. This is done `ReplLogStore::end_of_append_batch()`

Now regarding changes to be done:
This reminds me a change, I was thinking of doing, which is instead of relying on current state of Replica Set from RAFT (if its a leader or follower), I will track this in request structure itself (to make this decision of leader/follower). This way we don't need to worry about role changing in the middle. Will make that change in next commit.

Reply:
Relying on `read_sync` to ensure flush is done may not the right approach. Like you said our implementation of `entry_at` or even `read` without flush is on cards, so we must do a flush which is a no-op in those cases.